File: scala/src/main/scala/nl/quintor/akkastreams/StreamingMQTT.scala
package nl.quintor.akkastreams

import akka.Done
import akka.actor.ActorSystem
import akka.stream.alpakka.mqtt.scaladsl.MqttSource
import akka.stream.alpakka.mqtt.{MqttConnectionSettings, MqttMessage, MqttQoS, MqttSubscriptions}
import akka.stream.scaladsl.Source
import org.eclipse.paho.client.mqttv3.persist.MemoryPersistence

import scala.concurrent.Future

object StreamingMQTT extends App {
  implicit val system: ActorSystem = ActorSystem("Workshop")

  val connectionSettings = MqttConnectionSettings(
    "tcp://test.mosquitto.org:1883", // (1)
    "test-scala-client", // (2)
    new MemoryPersistence // (3)
  )

  val mqttSource: Source[MqttMessage, Future[Done]] =
    MqttSource.atMostOnce(
      connectionSettings,
      MqttSubscriptions("#", MqttQoS.atMostOnce),
      bufferSize = 8
    )
  mqttSource.map(message => (message.topic, message.payload.decodeString("UTF-8")))
    .filter(_._1.contains("/go-eCharger/"))
    .runForeach(Console.println)


}

File: java/src/main/java/nl/quintor/akkastreams/StreamingMQTT.java
package nl.quintor.akkastreams;

import akka.actor.ActorSystem;
import akka.stream.alpakka.mqtt.MqttConnectionSettings;
import akka.stream.alpakka.mqtt.MqttQoS;
import akka.stream.alpakka.mqtt.MqttSubscriptions;
import akka.stream.alpakka.mqtt.javadsl.MqttSource;
import org.eclipse.paho.client.mqttv3.persist.MemoryPersistence;
import scala.Tuple2;

public class StreamingMQTT {

  public static void main(String[] args) {
    final ActorSystem system = ActorSystem.create("QuickStart");

    MqttConnectionSettings connectionSettings = MqttConnectionSettings.create(
        "tcp://test.mosquitto.org:1883", // (1)
        "test-scala-client", // (2)
        new MemoryPersistence() // (3)
    );

    var mqttSource =
        MqttSource.atMostOnce(
            connectionSettings,
            MqttSubscriptions.create("#", MqttQoS.atMostOnce()),
            8
        );

    mqttSource.map(message -> new Tuple2<>(message.topic(), message.payload().decodeString("UTF-8")))
        .runForeach(tuple -> System.out.println(tuple._2), system);
  }

}
